### DISCUSSION TITLE
Speed on larger datasets.

### DISCUSSION THREAD

**Discussion by plezploz** [umami-software/umami#1926] [General]
I've noticed the application lag quite a lot when loading up all time statistics for websites (ex. umami's own demo https://app.umami.is/share/8rmHaheU/umami.is). Could this be possibly solved by adding some running counters for large sections of data?

**Comment by franciscao633** [comment]
Our cloud product uses Clickhouse, which doesn't have issues with the large datasets. However we are aware of performance issues in our self-hosting options (postgres/mysql) with larger datasets and explored options like TimescaleDB. We are open to other solutions as well and trying to balance performance with having as little dependencies to run Umami as well.

**Reply by icexuick** [reply]
Is there some info on what hardware a proper server needs to run Umami well on larger datasets.
We have some trackers with over 1M pageviews per month, and we can only use 7 days overviews - 30 days and higher crashes the server.

We use PHP/sql, currently on a 1 core 2GB ram server.
Should we upgrade to 2 cores/4GB or even 4 cores/8GB ?

**Reply by Matthijz98** [reply]
Can the self hosted version also use clickhouse?
Or do you just use the Postgres connector from clickhouse?

**Reply by franciscao633** [reply]
Currently not. Need to make some updates to write directly to CH, roll out some docs, setup, etc. The code is like 90% there to self-host a CH version of Umami.

**Reply by marcfowler** [reply]
I'd be very keen to see the Clickhouse version. Is there anything we can do to help get that over the line?

**Reply by wc-matteo** [reply]
same

**Reply by otahirs** [reply]
looking forward to CH support :pray:

**Reply by bgervan** [reply]
Is there any update on timescaledb? It would solve most of the issues for a single tenant (CH is more reasonable for the cloud multi tenant approach, but would be overkill for most self hoster)

**Comment by icexuick** [comment]
Does anybody have more information on better performance on self-hosting?
We have a site which gets between 1-2M pageviews per month, but we can only use the "7 day" overview. On "30 days" most of the data/graphs don't load anymore, and it crashes on "90 days".

We have tried cloud.umami for a month and a half, with over 1.3M page views, and that loads super-fast / instantly.

Any help / feedback / tips on how to get this working on a self-hosted solution is very much welcome!

**Comment by CammerJack** [comment]
what is considered large dataset? When it will start slowing down?

**Reply by jeffdrumgod** [reply]
My website_event table has over 19 million rows. The initial dashboard screen took more than 2 minutes to load and exceeded my system’s limits for the site with the highest number of records. Using EXPLAIN on the simple aggregation query, I found that over 251,000 rows were being examined.

(file related: src/queries/sql/getWebsiteStats.ts) 

I conducted a test by analyzing the query and noticed several existing indexes, but a key one was missing. I created the following index, and now the screen loads instantly in under 2 seconds:

```
CREATE INDEX idx_website_event_opt ON website_event (website_id, event_type, created_at, session_id, visit_id);
```

I plan to continue studying and propose improvements to the indexes in the Prisma schema in the coming days if I can find some time.

**Reply by otahirs** [reply]
great find 🚀. I will be waiting for more insights 👏
